mmv1/products/containerattached/terraform.yaml

```diff
@@ -0,0 +1,64 @@
+# Copyright 2022 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+--- !ruby/object:Provider::Terraform::Config
+overrides: !ruby/object:Overrides::ResourceOverrides
+  Cluster: !ruby/object:Overrides::Terraform::ResourceOverride
+    # Don't generate the LRO handling code because it doesn't account for regional services.
+    autogen_async: false
+    id_format: "projects/{{project}}/locations/{{location}}/attachedClusters/{{name}}"
+    import_format: ["projects/{{project}}/locations/{{location}}/attachedClusters/{{name}}"]
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "container_attached_cluster_minimal"
+        primary_resource_name: "fmt.Sprintf(\"basic%s\", context[\"random_suffix\"])"
+        primary_resource_id: "primary"
+        vars:
+          name: "basic"
+      - !ruby/object:Provider::Terraform::Examples
+        name: "container_attached_cluster_full"
+        primary_resource_name: "fmt.Sprintf(\"basic%s\", context[\"random_suffix\"])"
+        primary_resource_id: "primary"
+        vars:
+          name: "basic"
+    properties:
+      # The proto format of the authorization field is cumbersome to express in TF, so use a custom expander
+      # and flattener to massage the data.
+      authorization: !ruby/object:Overrides::Terraform::PropertyOverride
+        custom_expand: 'templates/terraform/custom_expand/containerattached_cluster_authorization_user.go.erb'
+        custom_flatten: 'templates/terraform/custom_flatten/containerattached_cluster_authorization_user.go.erb'
+      # Convenience expander to prepend "projects/" to fleet project numbers if not already present.
+      # Custom diff to take this convenience into account.
+      fleet.project: !ruby/object:Overrides::Terraform::PropertyOverride
+        diff_suppress_func: suppressAttachedClustersFleetProjectDiff
+        custom_expand: 'templates/terraform/custom_expand/containerattached_cluster_fleet_project.go.erb'
+      # If the user doesn't specify a loggingConfig, the server will supply a default value.
+      # No diff suppression because it is still possible to explicitly clear the config by passing
+      # an empty value.
+      loggingConfig: !ruby/object:Overrides::Terraform::PropertyOverride
+        default_from_api: true
```

Comment:
It might be worth double-checking that this behaves as expected; I'm not sure we have resources that use send_empty_value + allow_empty_object _and also_ default_from_api together yet.

Reply:
The basic test covers getting a default value from the service, the full test covers setting a value explicitly, and the (newly added) update test covers setting and then completely clearing the value. I think that's all the important scenarios? Is there something else I missed?

Reply:
hmm. so, thinking through it some more: the thing I'm concerned about is that this field seems to have the following potential values when you send it to the API:

1. Not set: API does nothing; values of this field are unmanaged by Terraform but will be stored in state.
2. Empty object: API clears field.
3. Non-empty object: API sets values according to object.

The weird thing here is that if you start with a non-empty object and then switch to "not set", the values for this field would (I think) be persisted in state indefinitely while not being present in the config.

Even though that's technically not an error, it seems like a pretty confusing / leaky user experience. I would expect that when I delete things from the config, they also get deleted in the API.

I think we should probably just not have this be a default_from_api field in Terraform - that is, make it required to send on every request. The problem is, that would require users to explicitly specify the field on every request - also not great. Maybe it could be marked optional but have a default value of an empty object that gets sent if the field is not set?

What do you think?

Reply:
On second thought, your original suggestion of sending an empty object when the user doesn't set anything does sound better. I'll do that.